<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns="http://www.w3.org/TR/REC-html40">

<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 9">
<meta name=Originator content="Microsoft Word 9">
<link rel=File-List href="./LutefiskXP%20Operators%20Manual_files/filelist.xml">
<link rel=Edit-Time-Data
href="./LutefiskXP%20Operators%20Manual_files/editdata.mso">
<!--[if !mso]>
<style>
v\:* {behavior:url(#default#VML);}
o\:* {behavior:url(#default#VML);}
w\:* {behavior:url(#default#VML);}
.shape {behavior:url(#default#VML);}
</style>
<![endif]-->
<title>Lutefisk1900 Operators Manual</title>
<style>
<!--
 /* Font Definitions */
@font-face
	{font-family:Ariel;
	panose-1:0 0 0 0 0 0 0 0 0 0;
	mso-font-alt:"Times New Roman";
	mso-font-charset:0;
	mso-generic-font-family:roman;
	mso-font-format:other;
	mso-font-pitch:auto;
	mso-font-signature:0 0 0 0 0 0;}
 /* Style Definitions */
p.MsoNormal, li.MsoNormal, div.MsoNormal
	{mso-style-parent:"";
	margin:0in;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
a:link, span.MsoHyperlink
	{color:blue;
	text-decoration:underline;
	text-underline:single;}
a:visited, span.MsoHyperlinkFollowed
	{color:purple;
	text-decoration:underline;
	text-underline:single;}
p
	{margin-right:0in;
	mso-margin-top-alt:auto;
	mso-margin-bottom-alt:auto;
	margin-left:0in;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
pre
	{margin:0in;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Courier New";}
@page Section1
	{size:8.5in 11.0in;
	margin:1.0in 1.25in 1.0in 1.25in;
	mso-header-margin:.5in;
	mso-footer-margin:.5in;
	mso-paper-source:0;}
div.Section1
	{page:Section1;}
 /* List Definitions */
@list l0
	{mso-list-id:15085927;
	mso-list-type:hybrid;
	mso-list-template-ids:-237612950 2039248262 -217800170 -1108325848 1848287280 -871839308 -1784241942 -619661832 1538010304 882388906;}
@list l1
	{mso-list-id:80950547;
	mso-list-type:hybrid;
	mso-list-template-ids:-925084546 -44422084 -2049813480 -1990534858 -2136934320 -1067695286 -104185678 1647486436 -943915058 690275496;}
@list l2
	{mso-list-id:117188209;
	mso-list-type:hybrid;
	mso-list-template-ids:213014886 11194850 1599536888 -145198248 1687571000 -1404268604 -211784132 1535540760 -802379522 -1961617792;}
@list l2:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l3
	{mso-list-id:122164918;
	mso-list-type:hybrid;
	mso-list-template-ids:-476428200 -854021566 419606746 197825072 -1035327746 1391464486 -845227592 -351641218 -1243461034 943888882;}
@list l3:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l4
	{mso-list-id:437608260;
	mso-list-type:hybrid;
	mso-list-template-ids:-2135239246 67698689 -905907426 67698693 67698689 67698691 67698693 67698689 67698691 67698693;}
@list l4:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	font-family:Symbol;}
@list l4:level2
	{mso-level-start-at:0;
	mso-level-number-format:bullet;
	mso-level-text:-;
	mso-level-tab-stop:1.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
@list l5
	{mso-list-id:638610202;
	mso-list-type:hybrid;
	mso-list-template-ids:1082668482 -520600158 -944980066 -1813230668 -1565866444 1398947310 759734578 -1545723752 -376293642 -821254990;}
@list l5:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l6
	{mso-list-id:691734351;
	mso-list-type:hybrid;
	mso-list-template-ids:1138230494 801133104 2043565212 1705834882 1978802024 476578118 1343907808 1756499626 177248442 -1189194142;}
@list l6:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l7
	{mso-list-id:792552912;
	mso-list-type:hybrid;
	mso-list-template-ids:1718789914 67698689 67698691 67698693 67698689 67698691 67698693 67698689 67698691 67698693;}
@list l7:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	font-family:Symbol;}
@list l8
	{mso-list-id:833838353;
	mso-list-type:hybrid;
	mso-list-template-ids:723807684 67698689 67698691 67698693 67698689 67698691 67698693 67698689 67698691 67698693;}
@list l8:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	font-family:Symbol;}
@list l9
	{mso-list-id:942034030;
	mso-list-type:hybrid;
	mso-list-template-ids:-1072024578 -16514398 2077024610 691823634 -2108939808 295195678 -1732744406 -1450835202 224958352 227287158;}
@list l9:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l10
	{mso-list-id:1014309733;
	mso-list-type:hybrid;
	mso-list-template-ids:155059720 -805289762 402421350 -1830502492 1954831836 1285167936 -399892482 -1371899156 183959302 559611712;}
@list l10:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l11
	{mso-list-id:1017079573;
	mso-list-type:hybrid;
	mso-list-template-ids:-681952902 1775531550 -738850632 -101786146 1303675942 586342036 1484582202 -1910593604 -2083740902 -1942745056;}
@list l11:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l12
	{mso-list-id:1083648824;
	mso-list-type:hybrid;
	mso-list-template-ids:1552593288 477267358 1741995834 -1674703114 975101374 2126959876 -1457377362 2104293834 1338433654 -1752168744;}
@list l13
	{mso-list-id:1098451769;
	mso-list-type:hybrid;
	mso-list-template-ids:1912904280 -1046740716 -998861166 -884937890 968017422 -2143634890 879368120 1069941018 -1125595974 -888242820;}
@list l13:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l14
	{mso-list-id:1163082219;
	mso-list-type:hybrid;
	mso-list-template-ids:1000104590 973350950 -1611340570 -747716254 1845524368 1018357452 -809307718 -220275180 887684204 1433704598;}
@list l14:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l15
	{mso-list-id:1520970387;
	mso-list-type:hybrid;
	mso-list-template-ids:118892200 67698689 67698691 67698693 67698689 67698691 67698693 67698689 67698691 67698693;}
@list l15:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	font-family:Symbol;}
@list l16
	{mso-list-id:1619025178;
	mso-list-type:hybrid;
	mso-list-template-ids:95992888 987766348 -1414617074 -1027468566 -206163838 1004565808 275697862 -1210552172 1216792744 1438570142;}
@list l16:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l17
	{mso-list-id:1638607686;
	mso-list-type:hybrid;
	mso-list-template-ids:-1990451576 -1932094682 -1418838898 1856926706 -2121748862 601243414 -1525613742 1122656050 1574705286 776919812;}
@list l17:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l18
	{mso-list-id:1653481060;
	mso-list-type:hybrid;
	mso-list-template-ids:545668174 67698689 67698691 67698693 67698689 67698691 67698693 67698689 67698691 67698693;}
@list l18:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	font-family:Symbol;}
@list l19
	{mso-list-id:1769812867;
	mso-list-type:hybrid;
	mso-list-template-ids:-1430731480 389701580 -1390483204 -918391538 456151340 -1908908920 933888750 -1686502682 -2113888204 1807758054;}
@list l19:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l20
	{mso-list-id:1797143061;
	mso-list-type:hybrid;
	mso-list-template-ids:-509432348 -2099232534 -654423674 -68098624 -2004478558 1867261782 -2046892698 840746494 1080196816 2120108212;}
@list l20:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l21
	{mso-list-id:1847594764;
	mso-list-type:hybrid;
	mso-list-template-ids:-295040522 1871102612 1194349460 -1351558860 739680960 1313624616 1397256062 -1274609670 -1511200924 2121801210;}
@list l21:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l22
	{mso-list-id:1861578459;
	mso-list-type:hybrid;
	mso-list-template-ids:845308870 812683814 1381136150 -204941920 -746414094 1870269736 -63247376 133309282 -1181875984 1244851706;}
@list l22:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l23
	{mso-list-id:1956591339;
	mso-list-type:hybrid;
	mso-list-template-ids:415675714 796189816 -1849230326 -1821623160 -1071480238 407910412 -1646104636 298741962 1948051248 -1185408746;}
@list l23:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l24
	{mso-list-id:2008169815;
	mso-list-type:hybrid;
	mso-list-template-ids:-40494772 -147585074 -1101000276 2055502988 -1629310182 1490296984 -208491248 -597154890 2052597602 1757026620;}
@list l25
	{mso-list-id:2012290201;
	mso-list-type:hybrid;
	mso-list-template-ids:-1425091878 -1326812384 764970520 -1283710348 113031260 1159113382 1233668100 787481554 -275234068 -360566924;}
@list l25:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l26
	{mso-list-id:2076002261;
	mso-list-type:hybrid;
	mso-list-template-ids:95614738 744008688 -988005584 -68254104 -1026397178 2059202448 460084772 666918296 218269314 -1727504420;}
@list l26:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
ol
	{margin-bottom:0in;}
ul
	{margin-bottom:0in;}
-->
</style>
<!--[if gte mso 9]><xml>
 <o:shapedefaults v:ext="edit" spidmax="1027"/>
</xml><![endif]--><!--[if gte mso 9]><xml>
 <o:shapelayout v:ext="edit">
  <o:idmap v:ext="edit" data="1"/>
 </o:shapelayout></xml><![endif]-->
<meta name=Template content="C:\Program Files\Microsoft Office\Office\html.dot">
</head>

<body bgcolor=white lang=EN-US link=blue vlink=purple style='tab-interval:.5in'>

<div class=Section1>

<p align=center style='text-align:center'>

<!doctype HTML>

<b><span style='font-size:24.0pt;font-family:Ariel'><!-- saved from url=(0066)http://helium.immunex.com/lutefisk/manual/Lutefisk1900_manual.html -->LutefiskXP
v1.0.5 Operators Manual</span></b> </p>


<div class=MsoNormal align=center style='text-align:center'>

<hr size=2 width="50%" align=center>

</div>


<div class=MsoNormal align=center style='text-align:center'>

<hr size=2 width="25%" align=center>

</div>


<p align=center style='text-align:center'><br>
<i>Documentation updated Aug. 26, 2005<br>
<br>
Copyright © 2005 Rich Johnson (jsrichar@alum.mit.edu)<br>
All rights reserved worldwide</i><br>
<br>
<br>
<br>
Lutefisk Homepage: <a href="http://www.sourceforge.net/"><span
style="mso-spacerun: yes"> </span>lutefiskxp.sourceforge.net<span
style='color:windowtext;text-decoration:none;text-underline:none'> </span></a></p>

<div align=center>

<table border=1 cellspacing=1 cellpadding=0 width=200 style='width:150.0pt;
 mso-cellspacing:.7pt;mso-padding-alt:0in 0in 0in 0in'>
 <tr>
  <td style='background:silver;padding:.75pt .75pt .75pt .75pt'>
  <p align=center style='text-align:center'><b>Table of Contents</b></p>
  </td>
 </tr>
 <tr>
  <td style='padding:.75pt .75pt .75pt .75pt'>
  <ol start=1 type=1>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l1 level1 lfo3;tab-stops:list .5in'><a href="#overview">Overview</a>
       </li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l1 level1 lfo3;tab-stops:list .5in'><a href="#files">Lutefisk
       files</a> </li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l1 level1 lfo3;tab-stops:list .5in'><a href="#compile">Compilation
       notes</a> </li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l1 level1 lfo3;tab-stops:list .5in'><a href="#contact">Contact
       information</a> </li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l1 level1 lfo3;tab-stops:list .5in'><a href="#version">Version
       history</a> </li>
  </ol>
  </td>
 </tr>
</table>

</div>


<div class=MsoNormal align=center style='text-align:center'><a name=overview>

<hr size=2 width="100%" align=center>

</a></div>


<p><span style='mso-bookmark:overview'><b><span style='font-size:14.0pt'>Overview</span></b>
</span></p>

<p><span style='mso-bookmark:overview'>Lutefisk is a program for the <i>de novo</i>
interpretation of peptide CID spectra. While it has a rudimentary interface, it
can be compiled for virtually any operating system with a C compiler. [Source
code and instructions are provided for MacOS, Win32, OSF, Solaris, Irix, and
Linux.].<span style="mso-spacerun: yes">  </span></span></p>

<p><span style='mso-bookmark:overview'>Lutefisk can be used in conjunction with
homology-based database search programs (e.g., OpenSea, MS-BLAST, or CIDentify)
as a supplement to standard MS/MS database search programs.<span
style="mso-spacerun: yes">  </span>We use it to find modified peptides or
sequence variants, as well as to help validate the results obtained from
database search programs (e.g, Mascot, Sequest, etc).<span style="mso-spacerun:
yes">  </span>We also find it useful for flagging high quality data that the
search programs (Mascot, etc) fail to identify, which may be due to a variety
of reasons (searching a human protein sequence database for an MSMS spectrum of
a mycobacterial derived peptide, for example).</span></p>

<span style='mso-bookmark:overview'></span>

<div class=MsoNormal align=center style='text-align:center'><a name=files>

<hr size=2 width="100%" align=center>

</a></div>


<p><span style='mso-bookmark:files'><b><span style='font-size:14.0pt'>Lutefisk
Files</span></b> </span></p>

<p><span style='mso-bookmark:files'>To run Lutefisk, you need to have four
files within the same directory or folder: </span></p>

<span style='mso-bookmark:files'></span>

<ol start=1 type=1>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l24 level1 lfo6;tab-stops:list .5in'><a href="#dataFile">CID data
     file</a> (data files can be specified with a full or partial pathname) </li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l24 level1 lfo6;tab-stops:list .5in'><a href="#detailFile">Lutefisk.details</a>
     </li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l24 level1 lfo6;tab-stops:list .5in'><a href="#paramsFile">Lutefisk.params</a>
     </li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l24 level1 lfo6;tab-stops:list .5in'><a href="#residuesFile">Lutefisk.residues</a>
     </li>
</ol>

<p>One additional file is optional: </p>

<ol start=1 type=1>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l12 level1 lfo9;tab-stops:list .5in'><a href="#databaseFile">Database.sequence</a>
     </li>
</ol>

<p>Once these files (containing the appropriate information as described below)
and the Lutefisk application are gathered together in one folder, you start the
application. Once execution begins, Lutefisk proceeds with minimal user
intervention. </p>

<p><i>On a Mac, a single dialog box appears where you can specify a variety of
command line arguments (if interested type -h to see help); in most cases, you
will click on the &quot;Ok&quot; button and use all of the default values. On
Windows, there is no such dialog box; however, command-line arguments can be
implemented by starting Lutefisk from the Command Prompt program supplied with
the Windows operating system. <o:p></o:p></i></p>

<p>A simulated teletype interface appears and indicates the various stages of
processing that have been achieved. When it is finished, the teletype interface
provides an initial list of sequences ranked in order of the &quot;intensity
score&quot;, followed by a more refined and shorter list of sequences. This
short list of sequences is also placed in a file with the default name
identical to the CID data file name with &quot;.lut&quot; appended. The header
to this file contains the information found in the parameter file
&quot;Lutefisk.params&quot;. </p>

<p><b>Note regarding use of Lutefisk with CIDentify:</b> The output file can be
read directly by the modified FASTA program called CIDentify. If you don't like
Lutefisk, you can still use CIDentify without using Lutefisk. This can easily
be done by editing a Lutefisk output file so that it contains your own
sequences (determined by hand or another sequencing program). Alternatively,
you can obtain the CIDentify source code and modify the data input format to
suit yourself. It is also worth pointing out that the Lutefisk output file can
be edited in order to eliminate any sequences that you somehow know are incorrect.
</p>

<p><a name=dataFile><b>CID data files<o:p></o:p></b></a></p>

<p><span style='mso-bookmark:dataFile'>Lutefisk can read CID data files in four
different formats: </span></p>

<ol start=1 type=1>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l0 level1 lfo12;tab-stops:list .5in'><span style='mso-bookmark:
     dataFile'>ASCII files created by the Finnigan TSQ program called
     &quot;List&quot;. These are created by starting the &quot;List&quot;
     program within ICIS Executive&quot; and opening the data file of interest
     within &quot;List&quot;. Under the &quot;File&quot; menu, go to
     &quot;Print...&quot;. A dialog box appears wherein you select
     &quot;ASCII&quot; as the saved formats, and under &quot;Text
     Displays&quot; select &quot;Multiple Pages&quot;. Provide a file name and
     select the &quot;Save to File&quot; button (don't select &quot;Print&quot;
     or else you will have reams of scratch paper). </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l0 level1 lfo12;tab-stops:list .5in'><span style='mso-bookmark:
     dataFile'>ASCII files created by the Finnigan LCQ File Converter program.
     From the destination box select &quot;text&quot; as the format, select the
     LCQ .raw files to convert, click on the arrow button, and then click on
     the convert button. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l0 level1 lfo12;tab-stops:list .5in'><span style='mso-bookmark:
     dataFile'>Tab-delineated ASCII text files. The first column contains m/z
     values followed by a tab and the second column is unitless relative
     intensity (an example is shown on the web site). </span></li>
</ol>

<p style='margin-left:.5in'><span style='mso-bookmark:dataFile'><i>For any of
these first three formats, the data can contain profile data with multiple data
points per mass unit, or it can be centroided (or peak top) m/z values. <o:p></o:p></i></span></p>

<ol start=4 type=1>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l0 level1 lfo12;tab-stops:list .5in'><span style='mso-bookmark:
     dataFile'>In addition, Lutefisk can read the Sequest &quot;.dta&quot;
     files; for &quot;.dta&quot; files &quot;C&quot; should be selected for the
     parameter &quot;Profile or Centroid&quot; found in the Lutefisk.params
     file (see below). This is our favorite data file format, since our
     Micromass Qtof and Thermofinnigan ion traps all have software that
     converts raw LC/MS/MS data to dta files. </span></li>
</ol>

<span style='mso-bookmark:dataFile'></span>

<p><a name=detailFile><b>Lutefisk.details<o:p></o:p></b></a></p>

<p><span style='mso-bookmark:detailFile'>The Lutefisk.details file contains the
so-called &quot;ion probabilities&quot; for each type of ion. Here is an </span><a
href="details.html">example</a>. Each column in the file contains the &quot;ion
probabilities&quot; for different fragmentation patterns (see the description
of &quot;fragmentation patterns&quot; below). Currently there are only two
types of fragmentation pattern that have been coded, which is for low energy
CID of tryptic peptides on triple quadrupole (or Qtof) instruments or ion
traps, and these ion probabilities are listed in the second and third columns.
The first column is not used (oddly enough).<span style="mso-spacerun: yes"> 
</span></p>

<p><a name=residuesFile><b>Lutefisk.residues<o:p></o:p></b></a></p>

<p><span style='mso-bookmark:residuesFile'>The Lutefisk.residues file contains
the single letter code, monoisotopic masses, average masses, and nominal masses
for each amino acid. The default Lutefisk.residue file is shown </span><a
href="residues.html">here</a>. To add an additional residue to the list,
replace the 0's in one of the rows w/ the corresponding monoisotopic, average,
and nominal masses. Up to five additional non-traditional residues can be
entered here, and will be given the single letter code of J, O, U, X, or Z.
Also, if you don't like my masses for the usual amino acids, you should feel
free to change them here. </p>

<p><a name=databaseFile><b>Database.sequence<o:p></o:p></b></a></p>

<p><span style='mso-bookmark:databaseFile'>The Database.sequence file is a text
file containing a sequence or a list of sequences that might have been derived
from a sequence database search. An example of such a file is shown </span><a
href="databaseSeq.html">here</a>. Although this seems like one is giving
Lutefisk the answers up front, in fact, Lutefisk will do its usual <i>de novo</i>
sequencing regardless of the Database.sequence list. In the final steps, where
it determines scores for the candidate sequences, Lutefisk tosses in these
database-derived sequences along with the <i>de novo</i> sequence candidates to
determine if the database sequences are as good as or better than the <i>de
novo</i> sequences. If so, then this constitutes evidence that the database
derived sequences might actually be correct. </p>

<p><a name=paramsFile><b>Lutefisk.params<o:p></o:p></b></a></p>

<p><span style='mso-bookmark:paramsFile'>The Lutefisk.params file is where most
of the user-selected variables are altered. Once appropriate parameters have
been chosen for a given set of data, one usually needs to change three
parameters -- &quot;CID Filename&quot;, &quot;Peptide MW&quot;, and
&quot;Charge-state&quot; (of the precursor ion). If the CID file is in the &quot;dta&quot;
format, the latter two parameters can be automatically read from the file
header and invoking the program like 'lutefisk <i>&lt;dta_file&gt;</i>' will
override all three parameters with those in the specifed data file(s). </span></p>

<p><span style='mso-bookmark:paramsFile'>Certain parameters need to be changed
to accommodate data obtained from different instruments. The mass tolerances
should match the anticipated errors for each type of instrument. The tolerance
parameter &quot;Final Fragment Err&quot; should be set to zero unless the data
was obtained from a Qtof, in which case, it should be a value of 0.02 - 0.05 (
I currently use 0.04). The &quot;Peak Width&quot; parameter should be set to 1
for unit resolved data (ion trap), and 0.75 for higher resolution data obtained
from a Qtof. Triple quads are often run in a low resolution mode to enhance
sensitivity, so the peak widths might be 2-3 u. For less than unit resolved
spectra (triple quad, say) set the &quot;Transition Mass&quot; to 1800. This is
the mass above which average (rather than monoisotopic) masses are used in the
calculations; for unit resolved or better data, set this high (5000) so that
average masses are never used. Since fragmentation patterns are slightly
different for triple quads, ion traps, and Qtof's, set the parameter
&quot;Fragmentation Pattern&quot; accordingly. Finally, the parameter
&quot;Auto Tag&quot; should be set to N for ion trap data, and Y for Qtof or
triple quad data. Here are the parameter files I use for data obtained by </span><a
href="LCMSTrap_Lutefisk.params">LC/MS/MS using an ion trap</a>, and <a
href="Qtof_Lutefisk.params">LC/MS/MS using a Qtof</a>. </p>

<p>Here is a more complete <a href="paramsDescrip.html">description of the
Lutefisk.params file parameters</a>. </p>

<p><a name=outputFile><b>Output files (.lut)<o:p></o:p></b></a></p>

<p><span style='mso-bookmark:outputFile'>The header repeats the information
contained in the params file, and also lists several scores that need some
explanation. The candidate sequences are ranked according to Pr(C) which is the
estimated probability of being correct.<span style="mso-spacerun: yes"> 
</span>I find that values over 0.5 are worth submitting to a homology-based
sequence database search program, and anything over 0.8 is particularly worthy
of serious consideration.<span style="mso-spacerun: yes">  </span>Pr(C) is
calculated from an empirically-derived 2-order polynomial fit to a weighted
average of the four remaining scores (Pevzscr, Quality, Intscr, and
X-corr).<span style="mso-spacerun: yes">  </span>Pevzscr is an adaptation of
the ideas presented by Dancik et al (J. of Comput. Biol (1999) Vol 6, 327),
which is a score that penalizes for the absence of expected ions and accounts
for the possibility of random matches.<span style="mso-spacerun: yes"> 
</span>Quality is the percentage of the peptide mass that can be accounted for
by a contiguous ion series.<span style="mso-spacerun: yes">  </span>Intscr is
the percentage of the fragment ion intensity that can be accounted for as b, y,
internal fragment, etc, ions.<span style="mso-spacerun: yes">  </span>X-corr is
the cross-correlation score that has been normalized by its auto-correlation
score.</span></p>

<p><span style='mso-bookmark:outputFile'>If &quot;Mass Scrambles for
Statistics&quot; in the params file was used, then the bottom of the output
file contains a summary of the statistical analysis. The first column &quot;1st
ranked&quot; lists the un-normalized scores for the top ranked sequence. The
column &quot;St Deviations&quot; shows how many standard deviations the top
ranked sequence scores were compared to the average wrong scores. The column
&quot;Average Wrong&quot; lists these wrong score averages, and the column
&quot;Correct/Wrong&quot; shows the ratio of the top correct score versus the
wrong score. Currently, I don’t recommend using this, so just give a zero value
for the parameter “Mass Scrambles for Statistics” (lutefisk.params file).</span></p>

<span style='mso-bookmark:outputFile'></span>

<div class=MsoNormal align=center style='text-align:center'><a name=compile>

<hr size=2 width="100%" align=center>

</a></div>


<p><span style='mso-bookmark:compile'><b><span style='font-size:13.5pt'>Compilation
Notes<o:p></o:p></span></b></span></p>

<p><span style='mso-bookmark:compile'>See the 'README' file for compilation
information.</span></p>

<p><span style='mso-bookmark:compile'><b>Unix/Linux:<o:p></o:p></b></span></p>

<p><span style='mso-bookmark:compile'>After untarring the archive, copy the makefile
for your system to &quot;Makefile&quot;. Use the &quot;make lutefisk&quot;
command.</span></p>

<p><span style='mso-bookmark:compile'><b>Win32:<o:p></o:p></b></span></p>

<p><span style='mso-bookmark:compile'>Current Metrowerks Projects are included
in the &quot;Win32&quot; folder. If you have an older compiler you will need to
create a new &quot;C Console App&quot; project and add the source files as specified
in the '0_ReadMe' file.</span></p>

<p><span style='mso-bookmark:compile'><b>Macintosh:<o:p></o:p></b></span></p>

<p><span style='mso-bookmark:compile'>Current Metrowerks Projects are included
in the &quot;Macintosh&quot; folder (provided as a self- extracting archive to
maintain fidelity). If you have an older compiler you will need to create a new
&quot;Std C Console PPC&quot; project and add the source files as specified in
the '0_ReadMe' file. Set the preferred heap size to 16M, the minimum heap size
to 8M, and the stack size to 512k. </span><a name=contact></a></p>


<div class=MsoNormal align=center style='text-align:center'><span
style='mso-bookmark:contact'>

<hr size=2 width="100%" align=center>

</span></div>


<p><span style='mso-bookmark:contact'><b><span style='font-size:13.5pt'>Contact
Information<o:p></o:p></span></b></span></p>

<span style='mso-bookmark:contact'></span>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><a
name=version>Lutefisk is software for de novo sequencing of peptides from
tandem mass spectra.</a></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>Copyright (C) 1995<span style="mso-spacerun:
yes">  </span>Richard S. Johnson</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>This program is free software; you can
redistribute it and/or</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>modify it under the terms of the GNU General
Public License</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>as published by the Free Software Foundation;
either version 2</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>of the License, or (at your option) any later
version.</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>This program is distributed in the hope that it
will be useful,</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>but WITHOUT ANY WARRANTY; without even the implied
warranty of</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.<span style="mso-spacerun: yes">  </span>See the</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>GNU General Public License for more details.</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>You should have received a copy of the GNU General
Public License</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>along with this program; if not, write to the Free
Software</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA<span style="mso-spacerun: yes">  </span>02110-1301, USA.</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>Contact:</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>Richard S Johnson</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>4650 Forest Ave SE</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>Mercer Island, WA 98040</span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p>

<p style='margin-bottom:12.0pt;tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>jsrichar@alum.mit.edu</span></p>


<div class=MsoNormal align=center style='text-align:center;tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'>

<hr size=2 width="100%" align=center>

</span></div>


<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b><span style='font-size:13.5pt'>Version History</span></b></span><span
style='mso-bookmark:version'><b><span style='font-size:10.0pt;mso-bidi-font-size:
13.5pt'><o:p></o:p></span></b></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'>LutefiskXP
version 1.0.4 – </span></b></span><span style='mso-bookmark:version'><i><span
style='mso-bidi-font-size:13.5pt'>External release</span></i></span><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'><o:p></o:p></span></b></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>When Q is in
the second position, then a c1 ion is present.<span style="mso-spacerun: yes"> 
</span>It is now scored, and used to distinguish the two amino acids if they
are not already.<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Limits placed
on the number of subsequences (drops by half) if the program has been
processing for over 30 seconds. It drops by another half if the processing is
over a minute. This will speed long ones up.<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Limits placed
on the deconvolution of the C-terminal unsequenced chunk of mass (in
Haggis).<span style="mso-spacerun: yes">  </span>If the overall processing has
gone on for over 45 sec then it drops the mass to 600.<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>For LCQ data,
b/y pairs are found and labeled as &quot;golden boys&quot; that cannot be
gotten rid of as easily.<span style="mso-spacerun: yes">  </span>These could
also be considered &quot;favored sons&quot;, although we can all hope that the
favored son &quot;W&quot; is eliminated in November (post-election news: sadly
the favored son is still in DC).<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Fixed the
ranking so that the db derived sequences don't fowl up the rank numbering.<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Fixed a bug
that, in certain cases, produced negative values for summed residue masses
displayed within brackets<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Changed the LCQ
b/y pair procedures (both in GetCID and main) such that pairs were either both
singly charged, or one was singly-charged and the other doubly-charged (but
only if the precursor is &gt; 2 charge state).<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Changed Haggis
a bit so that the minimum number of edges required to be considered a sequence
varied with the number of Lutefisk sequences already obtained as well as
peptide molecular weight.<span style="mso-spacerun: yes">  </span>Minimum
stayed at 4, but could be higher.<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Changed Haggis
so that if the array capacity was exceeded, it dumped the results obtained with
the Lutefisk sequences and then continued to run (rather than exit(1), which is
a waste).<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol;mso-bidi-font-weight:bold'>·<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Changed the way
the ion intensity is altered in the final scoring.<span style="mso-spacerun:
yes">  </span>The high intensity ions are reduced, but the low intensity ones
are not increased<b><o:p></o:p></b></span></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'>LutefiskXP
version 1.0 – </span></b></span><span style='mso-bookmark:version'><i><span
style='mso-bidi-font-size:13.5pt'>External release</span></i></span><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'><o:p></o:p></span></b></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>The Haggis
sequencing was modified so that two ion series could be combined.<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>The Haggis
sequencing was modified so that the two unsequenced masses at either end could
be matched to randomly<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>derived
sequences.<span style="mso-spacerun: yes">  </span>The random sequences that
fit with the most b and y ions is saved and replaces the chunk of mass.<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol'>·<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Added two output
variables to the params file.<span style="mso-spacerun: yes">  </span>Now can
specify the number of sequences and their Pr(c) limit.<o:p></o:p></span></span></p>

<p style='margin-left:.5in;text-indent:-.25in;mso-list:l8 level1 lfo15;
tab-stops:list .5in'><span style='mso-bookmark:version'><![if !supportLists]><span
style='mso-bidi-font-size:13.5pt;font-family:Symbol;mso-bidi-font-weight:bold'>·<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span><![endif]><span style='mso-bidi-font-size:13.5pt'>Modified final
scoring slightly, such that &quot;quality&quot; has less of an influence.<b><o:p></o:p></b></span></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'>Lutefisk1900
version 1.3.8 - </span></b><i>Internal release only<o:p></o:p></i></span></p>

<ul style='margin-top:0in' type=disc>
 <li class=MsoNormal style='mso-list:l4 level1 lfo18;tab-stops:list .5in'><span
     style='mso-bookmark:version'>Changed final scoring of candidate
     sequences.<span style="mso-spacerun: yes">  </span>About 150 spectra from
     both LCQ and Qtof were examined and sequence candidates with particularly
     high or low values for Pevscr (pavel pevzner), Intscr (intensity based
     scoring, quality (fraction of peptide mass accounted for by contiguous ion
     series) and xcorr (cross correlation score) were retained or
     discarded.<span style="mso-spacerun: yes">  </span>An average of these
     four scores was calculated and an emperically derived probability of being
     correct was determined.<span style="mso-spacerun: yes">  </span>The output
     is an estimated probability of being correct (Pr(c)).</span></li>
 <li class=MsoNormal style='mso-list:l4 level1 lfo18;tab-stops:list .5in'><span
     style='mso-bookmark:version'>Looks for pairs of y/b ions to re-determine
     the peptide MW in LCQ data.</span><span style='mso-bookmark:version'><b><span
     style='font-size:10.0pt;mso-bidi-font-size:13.5pt'><o:p></o:p></span></b></span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'>Lutefisk1900
version 1.3.7 - </span></b><i>Internal release only<o:p></o:p></i></span></p>

<ul style='margin-top:0in' type=disc>
 <li class=MsoNormal style='mso-list:l7 level1 lfo21;tab-stops:list .5in'><span
     style='mso-bookmark:version'>Made it so that the y1 and y2 goldenboy ions
     could not be removed on the basis of their intensity.</span></li>
 <li class=MsoNormal style='mso-list:l7 level1 lfo21;tab-stops:list .5in'><span
     style='mso-bookmark:version'>Added the Haggis type of subsequencing.<span
     style="mso-spacerun: yes">  </span>This finds series of ions connected by
     single amino acids, makes two sequences (forward and backward), and adds
     it to the list of sequence candidates.</span></li>
 <li class=MsoNormal style='mso-list:l7 level1 lfo21;tab-stops:list .5in'><span
     style='mso-bookmark:version'>Changed the &quot;fuzzy logic&quot; bit in
     LutefiskScorer so that mass variation from the calculated values have an
     affect on ion scores in a manner that follows a normal distribution
     (instead of a linear drop off).</span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'>Lutefisk1900
version 1.3.5 - </span></b><i>Internal release only<o:p></o:p></i></span></p>

<ul style='margin-top:0in' type=disc>
 <li class=MsoNormal style='mso-list:l15 level1 lfo24;tab-stops:list .5in'><span
     style='mso-bookmark:version'>Changed the way that the N- and C-terminal
     masses are entered.<span style="mso-spacerun: yes">  </span>Now it is
     possible to provide any mass rather than specific ones.</span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'>Lutefisk1900
version 1.3.4 - </span></b><i>Internal release only</i></span><span
style='mso-bookmark:version'><b><span style='mso-bidi-font-size:13.5pt'><o:p></o:p></span></b></span></p>

<ul style='margin-top:0in' type=disc>
 <li class=MsoNormal style='mso-list:l18 level1 lfo27;tab-stops:list .5in'><span
     style='mso-bookmark:version'>Fixed a buffer overflow problem in
     PeptideString() caused by long peptides.</span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1900 1.3.2 – </b><i>Released 1/28/02<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l3 level1 lfo30;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Revamped command-line invocation to be more intuitive. Now can
     say 'lutefisk *.dta' </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l3 level1 lfo30;tab-stops:list .5in'><span style='mso-bookmark:
     version'>The '-q' flag now completely quiets STDOUT output. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l3 level1 lfo30;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Fixed bug in GetCID so that plusArgLys array is initialized
     properly </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l3 level1 lfo30;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Set limits on &quot;quality&quot; and final combined scores,
     below which the sequences are trashed </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l3 level1 lfo30;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Minor change in final scoring of database-derived sequence </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l3 level1 lfo30;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Changed LutefiskScore so that when Qtof sequences are expanded,
     any given original sequence can only generate 250 new ones. When doing
     this procedure using the wrong peptide mass (statistical analysis) the
     program is limited to 50 new sequences for every old one. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l3 level1 lfo30;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Various minor bug fixes to aid compilation on some systems. </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1900 1.2.9 - </b><i>Internal release
only<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l10 level1 lfo33;tab-stops:list .5in'><span style='mso-bookmark:
     version'>The quality score from v1.2.8 is now used for screening candidate
     sequences. A minimum quality score is determined from the highest quality
     candidate, and candidates below this minimum are tossed out. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l10 level1 lfo33;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Various bug fixes.</span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1900 1.2.8 - </b><i>Internal release
only<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l21 level1 lfo36;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed the cross-correlation scoring so that instead of
     subtracting the average value of tau from -75 to 75, it does a point by
     point subtraction of the two numbers on each side of tau, takes the
     absolute value, and sums this over the range of -250 to 250. In other
     words, it subtracts tau(250) from tau(-250) and adds that to tau(249)-tau(-249),
     etc. This value is subtracted from tau(0) to give the cross-correlation
     score. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l21 level1 lfo36;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed the normalization of the cross-correlation score so that
     it is normalized to the auto-correlation of the spectrum itself (after the
     funny intensity normalizations). </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l21 level1 lfo36;tab-stops:list .5in'><span style='mso-bookmark:
     version'>The program can determine best scoring sequences for incorrect
     peptide molecular weights. It determines the best sequences for MW +/-
     14xN where N is any number from zero to 20 (or more if you want). It
     determines an average best wrong score and standard deviations around the
     mean, and then compares this with the best score for the correct peptide
     molecular weight. This allows for a statistical evaluation of the
     Lutefisk1900 results. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l21 level1 lfo36;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Because of all this looping, I found some memory leaks (fixed). </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l21 level1 lfo36;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Found a place where peptides exceeding the allowed length were
     stomping on some memory. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l21 level1 lfo36;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Added a spectrum quality assessment following completion of the
     sequencing. Its based on Pavel Pevzner comment that quality = #y or b ions
     / number of possible y or b ions. In practice, lutefisk will find the
     longest contiguous stretch of uninterrupted y or b ions and divide this by
     the (total number of amino acids minus one). A dipeptide in the sequence
     is not counted as an interrupted series; however, the dipeptide counts as
     two amino acids in the denominator. No skips are allowed.</span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1900 1.2.7 - </b><i>Internal release
only<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l5 level1 lfo39;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed ExpandSequence in LutefiskScore.c so that it checks to
     see how many new qtof sequences could be generated for each original
     sequence. In some rare instances (particularly for longer sequences), it
     was possible to expand a single sequence into many thousand related
     sequences (oxMet replacing Phe, or Gln replacing Lys, etc). Although not a
     bug, it was bogging things down. It now checks to make sure that there are
     no more than 500 new sequences expanded from the original; if there are
     too many, then it only allows sequence expansion for single amino acids
     (oxMet/Phe and Gln/Lys), and other reasons for expansion (multiple
     dipeptide choices, or Trp plus three dipeptides of the same mass) are
     eliminated. This should keep Lutefisk from getting hung up on certain
     spectra. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l5 level1 lfo39;tab-stops:list .5in'><span style='mso-bookmark:
     version'>checks to make sure that peptide lengths do not exceed the array
     limits -- if they do, then it just exits w/o going any further.</span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1900 1.2.6</b> - <i>Released 10/15/00<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l23 level1 lfo42;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed LutefiskGetCID so that it would not bomb-out when looking
     for a header to 'tab text' input file formats . </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l23 level1 lfo42;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed LutefiskScore so that it would include losses of CH3SOH from
     oxidized methionine. If the mass accuracy is sufficient to distinguish
     oxMet from Phe, then this loss is only considered for b and y ions that
     contain the oxMet. For lower accuracy, losses of 64 u from b and y ions
     that contain either Phe or oxMet are considered. </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1900 1.2.5</b> - <i>Internal release
only<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l19 level1 lfo45;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed LutefiskMakeGraph so that +2 y ions resulting from the
     loss of a single N-terminal amino acid is graphe d as a node even if there
     is no corresponding +1 y ion. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l19 level1 lfo45;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed main so that if the peakWidth parameter is set to zero
     for centroided data, that the autopeak width fea ture is not used. Its not
     possible to determine peakwidths from centroided data, and was getting
     slightly screwy results. If peakwidth is zero, then reasonable values are
     inserted depending on the instrument (0.75 for qtof a nd 1 for lcq). </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l19 level1 lfo45;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed LutefiskGetCID where there was an error in reading LCQ
     text files. The intensity values in the file ar e real, but were read into
     a struct value that needed an int; hence the intensity values were getting
     garbled. </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1900 1.2.4</b> - <i>Released 4/6/00<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l2 level1 lfo48;tab-stops:list .5in'><span style='mso-bookmark:
     version'>changed code so that data was not converted to nominal masses;
     this means that high mass accuracy measurements will not be wasted. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l2 level1 lfo48;tab-stops:list .5in'><span style='mso-bookmark:
     version'>because of temperature dependent drifts in the calibration slope
     of qtof data, the program will recalibrate the data for each candidate
     sequence. This allows for poorly calibrated qtof data to still have a
     tight fragment ion tolerance applied. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l2 level1 lfo48;tab-stops:list .5in'><span style='mso-bookmark:
     version'>the program will input sequences derived from database matches
     and compare these with the de novo derived candidate sequences. This
     allows for an electronic validation of a database hit. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l2 level1 lfo48;tab-stops:list .5in'><span style='mso-bookmark:
     version'>a CID data quality check has been added as an option. It looks
     for low mass amino acid related ions, high mass ions that can be connected
     by amino acid residue masses, as well as a general check on data
     strangeness (few ions, negative masses, precursor charge, etc). </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l2 level1 lfo48;tab-stops:list .5in'><span style='mso-bookmark:
     version'>code was modified to accomodate data that had been processed
     using Micromass's Maxent3 software. This software converts fragments to
     corresponding singly-charged values, and also removes isotope peaks. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l2 level1 lfo48;tab-stops:list .5in'><span style='mso-bookmark:
     version'>the lutefisk.residues file was added in order for people to add
     unusual amino acids to the list of common ones. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l2 level1 lfo48;tab-stops:list .5in'><span style='mso-bookmark:
     version'>some bugs were removed and others were added </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l2 level1 lfo48;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Compiled and run under MacOS, Win32, Solaris, OSF, SGI, &amp;
     Linux. </span></li>
</ul>

<p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
margin-left:.25in;tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 2.0.5</b> - <i>Released 10/16/98<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l13 level1 lfo51;tab-stops:list .5in'><span style='mso-bookmark:
     version'>bugs and the like </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 2.0.4<o:p></o:p></b></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l11 level1 lfo54;tab-stops:list .5in'><span style='mso-bookmark:
     version'>added stuff to make compatible with Codewarrior compilier for
     Win32 </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l11 level1 lfo54;tab-stops:list .5in'><span style='mso-bookmark:
     version'>added stuff to allow it to compile on DEC alpha </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 2.0.3<o:p></o:p></b></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l26 level1 lfo57;tab-stops:list .5in'><span style='mso-bookmark:
     version'>found a spot in GetCID where dividing by zero </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l26 level1 lfo57;tab-stops:list .5in'><span style='mso-bookmark:
     version'>added function to GetCID that eliminates ions below a specific
     s/n (SIGNAL_NOISE in the definitions file). The noise is determined in a
     100 da range local to each ion. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l26 level1 lfo57;tab-stops:list .5in'><span style='mso-bookmark:
     version'>added ability to read Sequest .dta files (Martin Baker) </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l26 level1 lfo57;tab-stops:list .5in'><span style='mso-bookmark:
     version'>added default output to match input name plus &quot;.lut&quot;
     (Martin Baker) </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 2.0.2<o:p></o:p></b></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l22 level1 lfo60;tab-stops:list .5in'><span style='mso-bookmark:
     version'>added ability to directly read Finnigan .dat files </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l22 level1 lfo60;tab-stops:list .5in'><span style='mso-bookmark:
     version'>rearranged the .params file into a more sensible order </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Verson 2.0.1<o:p></o:p></b></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l6 level1 lfo63;tab-stops:list .5in'><span style='mso-bookmark:
     version'>more bug fixes </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l6 level1 lfo63;tab-stops:list .5in'><span style='mso-bookmark:
     version'>modified to better handle ion trap data </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1.4.8<o:p></o:p></b></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l9 level1 lfo66;tab-stops:list .5in'><span style='mso-bookmark:
     version'>a few more bugs were found and fixed </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1.4.7<o:p></o:p></b></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l16 level1 lfo69;tab-stops:list .5in'><span style='mso-bookmark:
     version'>a few bugs that affected Unix operation were fixed </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1.4.6<o:p></o:p></b></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l25 level1 lfo72;tab-stops:list .5in'><span style='mso-bookmark:
     version'>auto-tag altered so that N-terminal mass can be three amino acids
     </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l25 level1 lfo72;tab-stops:list .5in'><span style='mso-bookmark:
     version'>in final scoring, sequences not ending in R or K for tryptic
     peptides are penalized </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l25 level1 lfo72;tab-stops:list .5in'><span style='mso-bookmark:
     version'>ions that connect to 147 or 175 for tryptic peptides were
     retained regardless of the intensity </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1.4.5<o:p></o:p></b></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l17 level1 lfo75;tab-stops:list .5in'><span style='mso-bookmark:
     version'>monoisotopic to average mass switch occurs over a 400 Da range </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l17 level1 lfo75;tab-stops:list .5in'><span style='mso-bookmark:
     version'>auto-tag automatically finds sequence tags prior to sequencing </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l17 level1 lfo75;tab-stops:list .5in'><span style='mso-bookmark:
     version'>auto-peakwidth automatically finds the peakwidth of profile data </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l17 level1 lfo75;tab-stops:list .5in'><span style='mso-bookmark:
     version'>for unit resolved spectra, the 2xC13 peak is discarded </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l17 level1 lfo75;tab-stops:list .5in'><span style='mso-bookmark:
     version'>cross-correlation bins every 0.5 Da instead of 1 Da </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l17 level1 lfo75;tab-stops:list .5in'><span style='mso-bookmark:
     version'>over-used ions in the final scoring are identified (same ions
     considered to be b and y) and final score reduced </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l17 level1 lfo75;tab-stops:list .5in'><span style='mso-bookmark:
     version'>the final scoring section was more carefully debugged and
     simplified </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1.3</b> - <i>Released 11/17/97<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l20 level1 lfo78;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Compilable with gcc 2.7 on UNIX. </span></li>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l20 level1 lfo78;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Command line options added to specify the output and param files.
     </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1.1</b> - <i>Released 6/22/97<o:p></o:p></i></span></p>

<ul type=disc>
 <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto;
     mso-list:l14 level1 lfo81;tab-stops:list .5in'><span style='mso-bookmark:
     version'>Fixed some problems in GetCidData() with reading of Finnigan
     ASCII files. </span></li>
</ul>

<p style='tab-stops:45.8pt 91.6pt 137.4pt 183.2pt 229.0pt 274.8pt 320.6pt 366.4pt 412.2pt 458.0pt 503.8pt 549.6pt 595.4pt 641.2pt 687.0pt 732.8pt'><span
style='mso-bookmark:version'><b>Version 1.0</b> - <i>Released 5/30/97</i> </span></p>

</div>

</body>

</html>
